docs/enhanced-provider-k8s.md

```diff
@@ -0,0 +1,303 @@
+
+# Managing Arbitrary Kubernetes Resources Using Crossplane Provider
+
+This document is aimed to explain how to manage arbitrary Kubernetes resources using Crossplane provider and how to enhance the provider by adding new features such as **resource management policy** and **resource reference** to manage Kubernetes resources more effectively.
+
+## About Provider Kubernetes
+
+The Crossplane Provider [provider-kubernetes](https://github.com/crossplane-contrib/provider-kubernetes) is designed to enable deployment and management of arbitrary Kubernetes resources in clusters. It defines an `Object` resource type which represents arbitrary Kubernetes resources, then uses a Kubernetes controller to reconcile `Object` typed resources to manage the corresponding Kubernetes resources.
+
+As an example, below YAML manifest defines an `Object` which represents a Kubernetes namesapce.
+
+```yaml
+apiVersion: kubernetes.crossplane.io/v1alpha1
+kind: Object
+metadata:
+  name: sample-namespace
+spec:
+  forProvider:
+    manifest:
+      apiVersion: v1
+      kind: Namespace
+      metadata:
+        # name in manifest is optional and defaults to Object name
+        # name: some-other-name
+        labels:
+          example: "true"
+  providerConfigRef:
+    name: kubernetes-provider
+```
+
+The detailed manifest for the namespace is defined in `spec.forProvider.manifest`. So, essentially, `Object` resource simply wraps the manifest of target Kubernetes resource, in this case the namespace. After `Object` resource is created, the provider will also watch the target Kubernetes resource and attach the runtime copy for the manifest including runtime status to the `Object` resource at `status.atProvider.manifest` so that we can monitor the target Kubernetes resource by checking the `Object` status field. 
+
+By convention in Crossplane, we usually call `Object` as managed resource, and the target Kubernetes resource as external resource. The managed resource `Object` is always co-located with the provider so that it can be watched by the provider. The external resource can be either co-located with the provider on the same cluster, or on a remote cluster, depending on the kubeconfig information defined in `spec.providerConfigRef` as above. Typically, the provider will check if the external resource exists on the target cluster, and if does not, it will create for us.
+
+![](images/provider-k8s-1.png)
+
+![](images/provider-k8s-2.png)
+
+## Why Provider Kubernetes
+
+When using Provider Kubernetes, you may think that it does not add too much value if it just wraps Kubernetes resource manifest as `Object` then does the transformation between the `Object` resource and the actual Kubernetes resource. This is because user can manage Kubernetes resource directly without using `Object` and the provider. They can manage Kubernetes resource either manually using command line tool such as kubectl or programmatically using normal Kubernetes controller.
```

Comment:
Ah, my bad, what I really mean here is that provider-kubernetes could bring more values to users if adding those new features :)

A little more reasoning as below for what I mentioned in this paragraph, correct me if I am wrong.

IMHO, provider-kubernetes has some unique characteristics than most of other providers. A Crossplane provider creates mapping from managed resources to external resources, where managed resources are always Kubernetes native resources handled by provider, and external resources can be anything in real world, e.g. in the case of AWS, it can be EC2, ELB, or EBS resource, etc.

Providers such as provider-aws bring us value to manage external resources via managed resources in Kubernetes way, no matter how they are handled in real world. On the other hand, provider-kubernetes is slightly different, since both managed and external resources are Kubernetes native resources. It might bring confusion to people who are new to this provider, e.g.:

If I have ways to manage my Kubernetes resources no matter using kubectl or whatever client in code, for remote provisioning, I can switch my kubeconfig from local one to remote one, why I have to turn the resources to `Object` with add additional layer using this provider, i.e.: to wrap one Kubernetes resource by another Kubernetes resource, unless it can give me additional values.

To avoid such confusion, I think it might be good to clarify clearly in provider-kubernetes README.

Reply:
I believe we are looking from different angles because my primary motivation with this provider was remote provisioning especially combined with Crossplane managed Kubernetes clusters (gke with provider-gcp), see the [in-composition example](https://github.com/crossplane-contrib/provider-kubernetes/blob/main/examples/in-composition/composition.yaml).

However, I see that you have some pretty valid scenarios with local k8s. I only have a slight concern that provider-kubernetes to turn into a backdoor to the Crossplane RBAC model. But given this requires an intentional step of adding kubeconfig of the local cluster with enough perms, this shouldn't be a blocker.

It would be great if you could soften the statements here considering remote provisioning case has its own uses cases and value :)